### PR TITLE
Feature ASCII encoding and decoding

### PR DESCRIPTION
**Pull request by jtenner**
Still lots of work to do

### CODE REVIEW

File: assembly/buffer/index.ts [hunk 1]
```diff
@@ -287,6 +287,33 @@ export class Buffer extends Uint8Array {
 }
 
 export namespace Buffer {
+  export namespace ASCII {
+    export function encode(str: string): ArrayBuffer {
+      let length = str.length;
+      let output = __alloc(length, idof<ArrayBuffer>());
+      for (let i = 0; i < length; i++) {
+        let char = load<u16>(changetype<usize>(str) + <usize>(i << 1));
+        store<u8>(output + <usize>i, <u8>(0b01111111 & char));
```
Comment (MaxGraey) [inline assembly/buffer/index.ts]: Could be simplify to
```suggestion
        store<u8>(output + <usize>i, char & 0x7F);
```

File: assembly/buffer/index.ts [hunk 2]
```diff
@@ -287,6 +287,33 @@ export class Buffer extends Uint8Array {
...
+    export function decode(buffer: ArrayBuffer): String {
+      return decodeUnsafe(changetype<usize>(buffer), buffer.byteLength);
+    }
+
+    @unsafe export function decodeUnsafe(pointer: usize, length: i32): String {
+      let result = __alloc(<usize>length << 1, idof<string>());
+
+      for (let i = 0; i < length; i++) {
+        let byte = load<u8>(pointer + <usize>i);
+        store<u16>(result + <usize>(i << 1), <u16>(byte & 0b01111111));
```
Comment (MaxGraey) [inline assembly/buffer/index.ts]:
```suggestion
        store<u16>(result + <usize>(i << 1), byte & 0x7F));
```

Comment (jtenner) [reply assembly/buffer/index.ts]: Thanks for this. I forgot that this might cause optimization problems with the upcasting.

**Comment by jtenner** [comment]
Anything else left for this?

**Comment by MaxGraey** [comment]
nope

**Pull request merged**